File: core/src/seal.ts
/**
 * sealBundle: compute canonical bytes, digest, set bundle_id, re-canonicalize,
 * re-digest, and produce a Seal with signature.
 */

import { canonicalBytes } from './canonical.js';
import { digest } from './digest.js';
import { sign, encodeBase64 } from './signature.js';
import type { Bundle, Seal, Header, SignatureAlgorithm, DigestAlgorithm } from './types.js';
import type { KeyPair } from './signature.js';

export class SealError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = 'SealError';
  }
}

export interface SealOptions {
  digestAlg?: DigestAlgorithm;
  sigAlg?: SignatureAlgorithm;
  id?: string;
}

export async function sealBundle(
  bundle: Bundle,
  keyPair: KeyPair,
  options: SealOptions = {}
): Promise<{ bundle: Bundle; seal: Seal }> {
  const digestAlg = options.digestAlg ?? 'SHA-256';
  const sigAlg = options.sigAlg ?? 'Ed25519';

  const hdr: Header = { ...bundle.hdr };
  if (options.id) {
    hdr.bundle_id = options.id;
  }

  const preBundle: Record<string, unknown> = {
    hdr,
    payload: bundle.payload,
  };

  if (bundle.meta !== undefined) {
    preBundle.meta = bundle.meta;
  }
  if (bundle.refs !== undefined && bundle.refs.length > 0) {
    preBundle.refs = bundle.refs;
  }

  const canon1 = canonicalBytes(preBundle);
  const digest1 = await digest(digestAlg, canon1);

  hdr.bundle_id = hdr.bundle_id ?? encodeBase64(digest1).replace(/[^a-zA-Z0-9_-]/g, '').slice(0, 32);

  const preBundle2: Record<string, unknown> = {
    hdr,
    payload: bundle.payload,
  };
  if (bundle.meta !== undefined) {
    preBundle2.meta = bundle.meta;
  }
  if (bundle.refs !== undefined && bundle.refs.length > 0) {
    preBundle2.refs = bundle.refs;
  }

  const canon2 = canonicalBytes(preBundle2);
  const digest2 = await digest(digestAlg, canon2);

  const signature = await sign(sigAlg, keyPair.privateKey, digest2);

  const seal: Seal = {
    digest_b64u: encodeBase64(digest2),
    digest_alg: digestAlg,
    sig_alg: sigAlg,
    signature_b64u: encodeBase64(signature),
  };

  return {
    bundle: {
      ...bundle,
      hdr,
      seal,
    },
    seal,
  };
}

File: python/proofbundle/seal.py
"""
Seal bundle: compute canonical bytes, digest, set bundle_id,
re-canonicalize, re-digest, and produce a seal with signature.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from .canonical import canonical_bytes
from .digest import digest
from .signature import KeyPair, encode_base64, export_key_pair


class SealError(Exception):
    pass


def _b64u(data: bytes) -> str:
    """Base64url encode without padding."""
    import base64
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def seal_bundle(
    bundle: Dict[str, Any],
    key_pair: KeyPair,
    digest_alg: str = "SHA-256",
    sig_alg: str = "Ed25519",
    bundle_id: Optional[str] = None,
) -> Dict[str, Any]:
    """Seal a bundle: compute digest and signature."""
    hdr = dict(bundle.get("hdr", {}))
    if bundle_id:
        hdr["bundle_id"] = bundle_id

    pre: Dict[str, Any] = {"hdr": hdr, "payload": bundle.get("payload")}
    meta = bundle.get("meta")
    if meta is not None:
        pre["meta"] = meta
    refs = bundle.get("refs")
    if refs is not None:
        pre["refs"] = refs

    canon1 = canonical_bytes(pre)
    digest1 = digest(digest_alg, canon1)

    if not bundle_id:
        hdr["bundle_id"] = _b64u(digest1)[:32]

    pre2: Dict[str, Any] = {"hdr": hdr, "payload": bundle.get("payload")}
    if meta is not None:
        pre2["meta"] = meta
    if refs is not None:
        pre2["refs"] = refs

    canon2 = canonical_bytes(pre2)
    digest2 = digest(digest_alg, canon2)

    from .signature import sign
    signature = sign(sig_alg, key_pair.private_key, digest2)

    # Export public key
    exported = export_key_pair(key_pair)
    pub_pem = exported["publicKey"]
    # Convert PEM to raw bytes for b64u encoding
    import base64
    pub_der = base64.b64decode(pub_pem.replace("-----BEGIN PUBLIC KEY-----", "")
                                .replace("-----END PUBLIC KEY-----", "")
                                .replace("\n", ""))

    seal = {
        "digest_alg": digest_alg,
        "digest_b64u": _b64u(digest2),
        "sig_alg": sig_alg,
        "signature_b64u": _b64u(signature),
    }

    result = dict(bundle)
    result["hdr"] = hdr
    result["seal"] = seal
    return result
